src/parquet/column_reader.cc

```diff
@@ -283,6 +286,405 @@ std::shared_ptr<ColumnReader> ColumnReader::Make(const ColumnDescriptor* descr,
+
+      records_read += DelimitRecords(num_records - records_read, &values_to_read);
+      reader->ConsumeBufferedValues(levels_read);
+    } else {
+      // No repetition or definition levels
+      batch_size = std::min(num_records - records_read, batch_size);
+      values_to_read = batch_size;
+      records_read += batch_size;
+      reader->ConsumeBufferedValues(values_to_read);
+    }
+
+    // Read any values before moving on to next data page
+    if (values_to_read > 0) {
+      ReadValues(reader, values_to_read, start_levels_position);
+    }
+  }
+
+  return records_read;
+}
+
+void RecordReader::ReadValues(ColumnReader* reader, const int64_t values_to_read,
+                              const int64_t start_levels_position) {
+  // Conservative upper bound
+  const int64_t possible_num_values =
+      std::max(values_to_read, levels_position_ - start_levels_position);
+  ReserveValues(possible_num_values);
+
+  int64_t null_count = 0;
+  if (nullable_values_) {
+    int64_t implied_values_read = 0;
+    internal::DefinitionLevelsToBitmap(
+        def_levels() + start_levels_position, levels_position_ - start_levels_position,
+        max_def_level_, max_rep_level_, &implied_values_read, &null_count,
+        valid_bits_->mutable_data(), values_written_);
+
+    ReadValuesSpaced(reader, values_to_read + null_count, null_count);
+  } else {
+    ReadValuesDense(reader, values_to_read);
+  }
+  // Total values, including null spaces, if any
+  values_written_ += values_to_read + null_count;
+  null_count_ += null_count;
+}
+
+int64_t RecordReader::DelimitRecords(int64_t num_records, int64_t* values_seen) {
+  int64_t values_to_read = 0;
+  int64_t records_read = 0;
+
+  const int16_t* def_levels = this->def_levels();
+  const int16_t* rep_levels = this->rep_levels();
+
+  if (max_rep_level_ > 0) {
+    // Count logical records and number of values to read
+    while (levels_position_ < levels_written_) {
+      if (rep_levels[levels_position_] == 0) {
+        at_record_start_ = true;
+        if (records_read == num_records) {
+          // We've found the number of records we were looking for
+          break;
+        } else {
+          // Continue
+          ++records_read;
+        }
+      } else {
+        at_record_start_ = false;
+      }
+      if (def_levels[levels_position_] == max_def_level_) {
+        ++values_to_read;
+      }
+      ++levels_position_;
+    }
+  } else {
+    if (max_def_level_ > 0) {
+      while (levels_position_ < levels_written_ &&
+             records_read < num_records) {
+        if (def_levels[levels_position_] == max_def_level_) {
+          ++values_to_read;
+        }
+        ++records_read;
+        ++levels_position_;
+      }
+    } else {
+      values_to_read = num_records;
+      records_read = num_records;
+    }
+  }
+  *values_seen = values_to_read;
+  return records_read;
+}
+
+void RecordReader::ReadValuesDense(ColumnReader* reader, int64_t values_to_read) {
```

Comment:
I will need to refactor these so the type-specific hot path can be inlined with no switch statements